cachi2/core/package_managers/npm.py

```diff
@@ -173,6 +185,115 @@ def get_sbom_components(self) -> list[dict[str, str]]:
+
+
+def _get_external_requirement_filepath(
+    download_path: RootedPath, package: str, requirement_url: str
+) -> Path:
+    if requirement_url.startswith("https://"):
+        algorithm = "sha512"
+        digest = "YOLO"
+        # TODO: Do I have to download that package first to get algorithm and digest???
+        filepath = Path(
+            download_path, f"external-{package}", f"{package}-external-{algorithm}-{digest}.tgz"
+        )
+    elif (
+        requirement_url.startswith("git+ssh://")
+        or requirement_url.startswith("github://")
+        or requirement_url.startswith("git+https://")
+    ):
+        parsed_url = urlparse(requirement_url)
+        namespace = parsed_url.path.removeprefix("/").split("/")[0]
+        repo = parsed_url.path.removeprefix("/").split("/")[-1].removesuffix(".git")
+        ref = parsed_url.fragment
+
+        filepath = Path(
+            download_path,
+            parsed_url.netloc,  # host
+            namespace,
+            repo,
+            f"{repo}-external-gitcommit-{ref}.tgz",
+        )
+    else:
+        raise ValueError(f"{requirement_url} is neither 'url' nor 'vcs'")
+    return filepath
+
+
+def clone_repo_pack_archive(
+    url: str,
+    download_path: Path,
+) -> None:
+    """
+    Clone a repository and pack its content as tar.
+
+    :param str url: URL for file download
+    :param str download_path: Path to download file to
+    :raise FetchError: If download failed
+    """
+    parse_url = url.split("#")
+    repo_url = parse_url[0]
+    repo_commit = parse_url[1] if len(parse_url) > 1 else None
+
+    # Create missing directories
+    directory = os.path.dirname(download_path)
+    os.makedirs(directory, exist_ok=True)
+
+    try:
+        clone_as_tarball(str(repo_url), str(repo_commit), download_path)
+    except FetchError:
+        log.info(f"Try to download git source with https:// {url}")
+        clone_as_tarball(
+            str(repo_url.replace("git+ssh://", "https://")), str(repo_commit), download_path
+        )
+
+
+def get_dependencies(download_dir: RootedPath, deps_to_download: Dict[str, Dict[str, Any]]) -> Dict:
+    """
+    Download npm dependencies.
+
+    Receives the url (proxy_repo_url), the destination directory (download_dir)
+    and the dependencies to be downloaded (deps_to_download).
+
+    :param Path download_dir: Path to download file to.
+    :param list[str] deps_to_download: Dict of dependencies to be downloaded.
+    :return: Dict of downloaded tarballs from npm registries for integrity check.
+    """
+    files_to_download = {}
+    for dependency_name, info in deps_to_download.items():
+        archive_name = f'{dependency_name}-{info["version"]}.tgz'.removeprefix("@").replace(
+            "/", "-"
+        )
+
+        if info["url"][-4:] != ".tgz":
+            clone_repo_pack_archive(
+                info["url"],
+                _get_external_requirement_filepath(download_dir, dependency_name, info["url"]),
+            )
```

Comment:
Can we do the "this is a git url" decision based on the scheme instead?

Plain https urls can end with things other than `.tgz` or even `.tar.gz` (such as `#<hash>` fragments)

Reply:
good point! I don't think we can use scheme to make right decision here. `https://` can be used for anything. So I tried to parse `file extension` from path using urlparse. How does this sound to you?

Reply:
To be a git url, the scheme must start with `git+` no? But checking for `tgz`, `tar.gz` should work too

Reply:
Yep your approach is better.